src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -16,36 +16,257 @@
+    Ok(Some(std::mem::replace(
+        &mut state.unit_dependencies,
+        HashMap::new(),
+    )))
+}
+
+/// Compute all the dependencies for libtest.
+/// Returns None if libtest is not needed.
+fn calc_deps_of_libtest<'a, 'cfg>(
+    mut state: &mut State<'a, 'cfg>,
+    std_roots: &[Unit<'a>],
+    roots: &[Unit<'a>],
+) -> CargoResult<Option<UnitGraph<'a>>> {
+    // Conditionally include libtest.
+    if std_roots.is_empty()
+        || !roots
+            .iter()
+            .any(|unit| unit.mode.is_rustc_test() && unit.target.harness())
+    {
+        return Ok(None);
+    }
+    state.is_std = true;
+    let test_id = state.resolve().query("test")?;
+    let test_pkg = state.get(test_id)?.expect("test doesn't need downloading");
+    let test_target = test_pkg
+        .targets()
+        .iter()
+        .find(|t| t.is_lib())
+        .expect("test has a lib");
+    let test_unit = new_unit(
+        state,
+        test_pkg,
+        test_target,
+        UnitFor::new_normal(),
+        Kind::Target,
+        CompileMode::Build,
+    );
+    let res = calc_deps_of_std(state, &[test_unit])?;
+    state.is_std = false;
+    Ok(res)
+}
+
+/// `proc_macro` has an implicit dependency on `std`, add it.
+fn fixup_proc_macro<'a>(std_unit_deps: &mut UnitGraph<'a>) {
+    // Synthesize a dependency from proc_macro -> std.
+    //
+    // This is a gross hack. This wouldn't be necessary with `--sysroot`. See
+    // also libtest below.
+    if let Some(std) = std_unit_deps
+        .keys()
+        .find(|unit| unit.pkg.name().as_str() == "std" && unit.target.is_lib())
+        .cloned()
+    {
+        for (unit, deps) in std_unit_deps.iter_mut() {
+            if unit.pkg.name().as_str() == "proc_macro" {
+                deps.push(UnitDep {
+                    unit: std,
+                    unit_for: UnitFor::new_normal(),
+                    extern_crate_name: InternedString::new("std"),
+                    public: true,
+                });
+            }
+        }
+    }
+}
+
+/// Add libtest as a dependency of any test unit that needs it.
+fn attach_std_test<'a, 'cfg>(
+    state: &mut State<'a, 'cfg>,
+    mut libtest_unit_deps: UnitGraph<'a>,
+    std_unit_deps: &UnitGraph<'a>,
+) {
+    // Attach libtest to any test unit.
+    let (test_unit, test_deps) = libtest_unit_deps
+        .iter_mut()
+        .find(|(k, _v)| k.pkg.name().as_str() == "test" && k.target.is_lib())
+        .expect("test in deps");
+    for (unit, deps) in state.unit_dependencies.iter_mut() {
+        if unit.kind == Kind::Target && unit.mode.is_rustc_test() && unit.target.harness() {
+            // `public` here will need to be driven by toml declaration.
+            deps.push(UnitDep {
+                unit: *test_unit,
+                unit_for: UnitFor::new_normal(),
+                extern_crate_name: test_unit.pkg.name(),
+                public: false,
+            });
+        }
+    }
+
+    // Synthesize a dependency from libtest -> libc.
```

Comment:
I'll think about this and experiment with it.  I'm a little uncertain how that would interact if the user has their own `libc` dependency, or if they don't, could they say `extern crate libc;` and circumvent the `rustc_private` check, or how it would behave.

Reply:
I *think* that the stability is handled alright since the entire crate is marked as unstable so usage of anything will trigger a privacy warning. Even `extern crate libc` would trigger (or so I think it will).

Reply:
I've been experimenting with this, and I'm finding it more difficult and complicated than the current solution. I keep needing to add more special logic for std, and I would prefer to go in the other direction.

This only works if `-Z force-unstable-if-unmarked` is passed to rustc. Should I add that flag to all sysroot crates? It will need an `is_std` flag in `Unit` to detect.

I'm concerned about the explosion of rustc args. This would go from ~6 extra `--extern` flags (`compiler_builtins`, `core`, `std`, `panic_unwind`, `std` (twice, for rlib/dylib)) to ~20 `--extern` flags.  This alone adds about 3k to the command-line length for me.

I'm not really convinced doing this buys a lot. It removes one hard-coded dependency (libtest→libc). It also doesn't help with proc_macro→std.

I've been trying to think of alternate ways to handle this. Some rough thoughts:

Perhaps in the future we can instead just make them explicit. Like libtest could have something like:
```toml
[dependencies]
libc = { sysroot=true }
```

Or maybe the way libtest/proc_macro are built could be changed. I've never really understood why they are built separately from libstd. If they were all built together and shared the same CARGO_TARGET directory, then they could have real dependency declarations (like libtest could say `libc = { version = "0.2.51", default-features = false, features = ['rustc-dep-of-std'] }` and it all would just work, no more implicit dependencies).

Dunno, just thinking out-loud. It just doesn't seem right to pass a bunch of internal std dependencies to every crate.

Reply:
Ok sounds like "pass `--extern` everywhere" is definitely not the strategy to go.

> Or maybe the way libtest/proc_macro are built could be changed.

This sounds like the way to go I think. The way we build libtest hasn't change since, well, ever. The original problem was that libtest wants to depend on crates.io crates (`getopts`) and it was thought that this would grow over time. This in turn means that libstd has to be built and in the sysroot because Cargo doesn't know that dependencies from crates.io otherwise depend on libstd (we could already of course change libtest-in-tree to do this). 

Nowadays though with the advent of `[patch]` and the workspace workarounds already present in rust-lang/rust, we've actually got libstd itself depending on crates.io crates. I suspect that a similar strategy is probably all we need to do to get libtest working as well. 

I would sort of like to improve how libstd depends on crates.io crates (and how libtest would as well) but for now I think we should probably just say "the best way to fix this part of Cargo is to make libtest/libproc_macro a full dependency graph with libstd".

I'll work on getting this upstream in libstd itself.

Reply:
Ok I've posted https://github.com/rust-lang/rust/pull/63637 which should obsolete this dependency drawing in theory.

That being said I think it's fine to go ahead and land this and we can just back it out later